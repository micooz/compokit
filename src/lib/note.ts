import { dedupBy } from "./common";
import { Inter, Interval } from "./interval";

export enum NoteEnum {
  C = 1,
  D,
  E,
  F,
  G,
  A,
  B,
}

export enum AccidentalEnum {
  Natural,
  NaturalWithSymbol,
  Sharp,
  Flat,
  DoubleSharp,
  DoubleFlat,
}

const abbrToNoteEnum: Record<string, NoteEnum> = {
  C: NoteEnum.C,
  D: NoteEnum.D,
  E: NoteEnum.E,
  F: NoteEnum.F,
  G: NoteEnum.G,
  A: NoteEnum.A,
  B: NoteEnum.B,
};

const abbrToAccidentalEnum: Record<string, AccidentalEnum> = {
  "": AccidentalEnum.Natural,
  "#": AccidentalEnum.Sharp,
  "##": AccidentalEnum.DoubleSharp,
  b: AccidentalEnum.Flat,
  bb: AccidentalEnum.DoubleFlat,

  "♮": AccidentalEnum.NaturalWithSymbol,
  "♯": AccidentalEnum.Sharp,
  "𝄪": AccidentalEnum.DoubleSharp,
  "♭": AccidentalEnum.Flat,
  "𝄫": AccidentalEnum.DoubleFlat,
};

const noteEnumToString: Record<NoteEnum, string> = {
  [NoteEnum.C]: "C",
  [NoteEnum.D]: "D",
  [NoteEnum.E]: "E",
  [NoteEnum.F]: "F",
  [NoteEnum.G]: "G",
  [NoteEnum.A]: "A",
  [NoteEnum.B]: "B",
};

const accidentalEnumToString: Record<AccidentalEnum, string> = {
  [AccidentalEnum.Natural]: "",
  [AccidentalEnum.NaturalWithSymbol]: "♮",
  [AccidentalEnum.Sharp]: "♯",
  [AccidentalEnum.Flat]: "♭",
  [AccidentalEnum.DoubleSharp]: "𝄪",
  [AccidentalEnum.DoubleFlat]: "𝄫",
};

const accidentalEnumToHumanString: Record<AccidentalEnum, string> = {
  [AccidentalEnum.Natural]: "",
  [AccidentalEnum.NaturalWithSymbol]: "",
  [AccidentalEnum.Sharp]: "#",
  [AccidentalEnum.Flat]: "b",
  [AccidentalEnum.DoubleSharp]: "##",
  [AccidentalEnum.DoubleFlat]: "bb",
};

const accidentalOffsetMap = {
  [AccidentalEnum.Natural]: 0,
  [AccidentalEnum.NaturalWithSymbol]: 0,
  [AccidentalEnum.Sharp]: 1,
  [AccidentalEnum.Flat]: -1,
  [AccidentalEnum.DoubleSharp]: 2,
  [AccidentalEnum.DoubleFlat]: -2,
};

const noteNames = Object.keys(abbrToNoteEnum);
const noteAbbrRegex = new RegExp(
  `^([${noteNames.join("")}])([^\\d]{0,})(\\d{0,})$`,
  "i"
);

export type NoteToNameOptions = {
  transformAccidental?: boolean;
};

export type NoteIsOptions = {
  checkGroup?: boolean;
  checkAccidental?: boolean;
};

export class Note {
  /**
   * C、D、E、F、G、A、B
   */
  private _note: NoteEnum;

  /**
   * 0-9
   * Default: 0
   */
  private _group?: number;

  /**
   * Default: AccidentalEnum.Natural
   */
  private _accidental: AccidentalEnum;

  constructor(note: NoteEnum, group?: number, accidental?: AccidentalEnum) {
    this._note = note;
    this._group = group;
    this._accidental = accidental ?? AccidentalEnum.Natural;
  }

  static from(src: NoteType) {
    if (src instanceof Note) {
      return src.clone();
    }

    const abbr = src;

    const matches = abbr.match(noteAbbrRegex);

    if (!matches) {
      throw new Error(`invalid abbr: ${abbr}`);
    }

    let [_, _note, _acc, _group] = matches;

    const note = abbrToNoteEnum[_note.toUpperCase()];

    // if (typeof note === "undefined") {
    //   throw new Error(`unknown note: ${_note}`);
    // }

    const accidental = abbrToAccidentalEnum[_acc];

    if (typeof accidental === "undefined") {
      throw new Error(`invalid accidental: ${_acc}`);
    }

    const group = _group === "" ? undefined : Number(_group);

    if (group && (group < 0 || group > 9)) {
      throw new Error(`invalid group: ${group}`);
    }

    return new Note(note, group, accidental);
  }

  static fromIndex(index: number) {
    if (index >= NoteEnum.C && index <= NoteEnum.B) {
      return new Note(index);
    }

    let final = index % 7;

    if (final <= 0) {
      final += 7;
    }

    return new Note(final);
  }

  get index() {
    return this._note;
  }

  get accidental() {
    return this._accidental;
  }

  set accidental(value: AccidentalEnum) {
    this._accidental = value;
  }

  get group() {
    return this._group;
  }

  set group(value: number | undefined) {
    this._group = value;
  }

  isNature() {
    return (
      this._accidental === AccidentalEnum.Natural ||
      this._accidental === AccidentalEnum.NaturalWithSymbol
    );
  }

  isSharp() {
    return this._accidental === AccidentalEnum.Sharp;
  }

  isFlat() {
    return this._accidental === AccidentalEnum.Flat;
  }

  clone() {
    return new Note(this._note, this._group, this._accidental);
  }

  name(opts?: NoteToNameOptions) {
    const { transformAccidental = false } = opts || {};

    const note = noteEnumToString[this._note];
    const accidental = transformAccidental
      ? accidentalEnumToString[this._accidental]
      : accidentalEnumToHumanString[this._accidental];

    return `${note}${accidental}`;
  }

  nameWithGroup(opts?: NoteToNameOptions) {
    let name = this.name(opts);

    if (typeof this._group !== "undefined") {
      name += this._group;
    }

    return name;
  }

  is(note: NoteType, opts?: NoteIsOptions) {
    const { checkGroup = false, checkAccidental = false } = opts || {};

    const _note = Note.from(note);
    const conditions: boolean[] = [];

    conditions.push(this.index === _note.index);

    if (checkGroup && (this.group !== undefined || _note.group !== undefined)) {
      conditions.push(this.group === _note.group);
    }
    if (checkAccidental && (!this.isNature() || !_note.isNature())) {
      conditions.push(this.accidental === _note.accidental);
    }

    return conditions.every((item) => item);
  }

  add(inter: Inter) {
    const interval = Interval.from(inter);

    if (interval.is("H1")) {
      throw new Error("unsupported interval: H1");
    }

    if (interval.is("P1")) {
      return this.clone();
    }

    const degree = interval.degree.valueOf();
    const nextNoteAbs = this._note.valueOf() + degree - 1;

    // move to the same octave
    const nextNote = Note.fromIndex(nextNoteAbs);

    // across how many octaves
    nextNote.group =
      this._group === undefined
        ? this._group
        : this._group + Math.floor((nextNoteAbs - 1) / 7);

    // decide next note's accidental
    const diff = interval.halfStepCount - this.to(nextNote).halfStepCount;

    const accidental = {
      [-2]: AccidentalEnum.DoubleFlat,
      [-1]: AccidentalEnum.Flat,
      [0]: AccidentalEnum.Natural,
      [12]: AccidentalEnum.Natural,
      [1]: AccidentalEnum.Sharp,
      [2]: AccidentalEnum.DoubleSharp,
    }[diff];

    nextNote.accidental = accidental!;

    return nextNote;
  }

  minus(inter: Inter) {
    const interval = Interval.from(inter);

    if (interval.is("H1")) {
      throw new Error("unsupported interval: H1");
    }

    if (interval.is("P1")) {
      return this.clone();
    }

    const degree = interval.degree.valueOf();
    const nextNoteAbs = this._note.valueOf() - degree + 1;

    // move to the same octave
    const nextNote = Note.fromIndex(nextNoteAbs);

    // across how many octaves
    nextNote.group =
      this._group === undefined
        ? this._group
        : this._group + Math.floor((nextNoteAbs - 1) / 7);

    // decide next note's accidental
    const diff = interval.halfStepCount - nextNote.to(this).halfStepCount;

    const accidental = {
      [2]: AccidentalEnum.DoubleFlat,
      [1]: AccidentalEnum.Flat,
      [0]: AccidentalEnum.Natural,
      [12]: AccidentalEnum.Natural,
      [-1]: AccidentalEnum.Sharp,
      [-2]: AccidentalEnum.DoubleSharp,
    }[diff];

    nextNote.accidental = accidental!;

    return nextNote;
  }

  to(note: Note) {
    // 1. calculate degree number
    let degree = note.index - this.index + 1;

    if (degree < 1) {
      degree += 7;
    }

    // 2. calculate half step count
    let halfStepCount = degree * 2 - 2;

    let from = this.index;

    // in white keys
    while (from !== note.index) {
      const note = Note.fromIndex(from + 1);

      if (note.is("F") || note.is("C")) {
        halfStepCount -= 1;
      }

      from = note.index;
    }

    // consider current note's accidental
    halfStepCount -= accidentalOffsetMap[this._accidental];

    // consider target note's accidental
    halfStepCount += accidentalOffsetMap[note.accidental];

    // in case C# to C is -1
    if (halfStepCount === -1) {
      halfStepCount = 1;
    }

    return Interval.fromDegreeAndHalfStepCount(degree, halfStepCount);
  }
}

export class MidiNote extends Note {
  /**
   * 0-127
   */
  private _velocity?: number;
}

export class NoteArray {
  private _notes: Note[] = [];

  constructor(notes: Notes) {
    if (notes.length === 0) {
      this._notes = [];
    }
    if (typeof notes[0] === "string") {
      this._notes = (notes as string[]).map((note) => Note.from(note));
    }
    if (notes[0] instanceof Note) {
      this._notes = (notes as Note[]).map((note) => note.clone());
    }
  }

  static from(notes: Notes) {
    return new NoteArray(notes);
  }

  valueOf() {
    return this._notes;
  }

  count() {
    return this._notes.length;
  }

  get(index: number) {
    return this._notes[index].clone();
  }

  root() {
    return this._notes[0].clone();
  }

  /**
   * calculate intervals in pairs
   */
  intervals() {
    const intervals: Interval[] = [];

    this.forPair((prev, next) => {
      const inter = prev.to(next);
      intervals.push(inter);
    });

    return intervals;
  }

  map(func: (note: Note, index: number, array: Note[]) => Note) {
    const notes = this._notes.map(func);
    return new NoteArray(notes);
  }

  dedup() {
    const notes = dedupBy(this._notes, (note) => note.name());
    return new NoteArray(notes);
  }

  forPair(func: (prev: Note, next: Note) => void) {
    return this._notes.reduce((prev, next) => {
      func(prev, next);
      return next;
    });
  }

  clone() {
    return new NoteArray(this._notes.map((note) => note.clone()));
  }

  join(separator = " ", opts?: NoteToNameOptions) {
    return this.names(opts).join(separator);
  }

  names(opts?: NoteToNameOptions) {
    return this._notes.map((note) =>
      note.nameWithGroup({ transformAccidental: true, ...opts })
    );
  }

  withGroup(from: number) {
    let group = from;

    const notes = this._notes.map((note, index) => {
      const currentNote = note.clone();
      const prevNote = this._notes[index - 1];

      if (prevNote && currentNote.index < prevNote.index) {
        group += 1;
      }

      currentNote.group = group;

      return currentNote;
    });

    return new NoteArray(notes);
  }

  include(note: Note, opts?: NoteIsOptions) {
    for (const item of this._notes) {
      if (note.is(item, opts)) {
        return true;
      }
    }
    return false;
  }

  omit(notes: NoteArray, opts?: NoteIsOptions) {
    const newNotes = this._notes.filter((item) => !notes.include(item));
    return NoteArray.from(newNotes);
  }
}

export type NoteType = string | Note;

export type Notes = string[] | Note[];
