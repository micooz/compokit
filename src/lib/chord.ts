import { Note, NoteToNameOptions, Notes } from "./note";
import { NoteArray } from "./note-array";

export enum ChordTypeEnum {
  Triad,
  Seventh,
  // Ninth,
  // Eleventh,
  // Thirteen,
}

const chordQualityMap: Record<string, string> = {
  "m-m": "dim",
  "m-M": "m",
  "M-m": "", // the triad doesn't need an "M"
  "M-M": "aug",

  "m-m-m": "dim7",
  "m-m-M": "m7(♭5)",
  "m-M-m": "m7",
  "m-M-M": "mM7",
  "M-m-m": "7",
  "M-m-M": "maj7",
  "M-M-m": "aug7",
  "M-M-M": "augM7",
};

const degreeSeqToInversionAndRootIndex: Record<string, [number, number]> = {
  "3-3": [0, 0], // C E G
  "3-4": [1, 2], // E G C
  "4-3": [2, 1], // G C E
  "3-3-3": [0, 0], // C E G B
  "3-3-2": [1, 3], // E G B C
  "3-2-3": [2, 2], // G B C E
  "2-3-3": [3, 1], // B C E G
};

export class Chord {
  private _noteArr: NoteArray;

  private _root: Note;

  private _inversion = 0;

  constructor(notes: Notes) {
    this._noteArr = new NoteArray(notes);

    // TODO: supports structures containing omitted and added notes
    if (notes.length < 3) {
      throw new Error("chord must contain at least three note");
    }

    // dedup
    const dedup = this._noteArr.dedup();

    // intervals in pairs
    const intervals = dedup.intervals();

    // extract degrees
    const degrees = intervals.map((inter) => inter.degree.valueOf());

    // determine inversion and root index
    const [inversion, rootIndex] =
      degreeSeqToInversionAndRootIndex[degrees.join("-")] || [];

    if (inversion === undefined || rootIndex === undefined) {
      throw new Error(
        `cannot form a chord with these notes: ${this._noteArr.join()}`
      );
    }

    // root note
    const root = dedup.get(rootIndex);

    this._root = root;
    this._inversion = inversion;
  }

  private intervals() {
    return this._noteArr.intervals();
  }

  // TODO: fromAbbr
  fromAbbr(abbr: string) {
    //
  }

  inversion(ordinal = 1) {
    const backward = ordinal < 0;
    const notes = this._noteArr.clone().valueOf();

    for (let n = 1; n <= Math.abs(ordinal); n += 1) {
      if (backward) {
        const last = notes.pop()!;

        if (last.group) {
          last.group -= 1;
        }

        notes.unshift(last);
      } else {
        const first = notes.shift()!;

        if (first.group) {
          first.group += 1;
        }

        notes.push(first);
      }
    }

    return new Chord(notes);
  }

  inversionOrdinal() {
    return this._inversion;
  }

  rootPosition() {
    if (this._inversion === 0) {
      return this.clone();
    }

    return this.inversion(-this._inversion);
  }

  toAbbr(opts?: NoteToNameOptions) {
    const first = this._noteArr.get(0).name(opts);

    // C
    const root = this._root.name(opts);

    // chord in root position
    const inRootPosition = this.inversion(-this._inversion);

    // quality sequence
    const qualitySeq = inRootPosition
      .intervals()
      .map((quality) => {
        if (quality.isMinor()) {
          return "m";
        }
        if (quality.isMajor()) {
          return "M";
        }
        if (quality.isAugmented()) {
          return "aug";
        }
        if (quality.isDiminished()) {
          return "dim";
        }
      })
      .filter(Boolean)
      .join("-");

    // m/M/aug/dim(7/9/11/13)
    const quality = chordQualityMap[qualitySeq];

    if (quality === undefined) {
      throw new Error(
        `unable to determine the appropriate chord representation for: ${this.join()}`
      );
    }

    // (♭5)(♭9)(♭13)
    const accidentals = [].join("");

    // /G
    let by = "";

    if (root !== first) {
      by = `/${first}`;
    }

    // CmM7/G
    return `${root}${quality}${accidentals}${by}`;
  }

  notes() {
    return this._noteArr;
  }

  noteNames(opts?: NoteToNameOptions) {
    return this._noteArr.names(opts);
  }

  join(separator = " ", opts?: NoteToNameOptions) {
    return this._noteArr.join(separator, opts);
  }

  clone() {
    return new Chord(this._noteArr.valueOf());
  }
}
