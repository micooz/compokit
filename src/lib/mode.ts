import { Chord, ChordTypeEnum } from "./chord";
import { transformObject } from "./common";
import { Inter, Interval } from "./interval";
import {
  AccidentalEnum,
  Note,
  NoteToNameOptions,
  NoteType,
  Notes,
} from "./note";
import { NoteArray } from "./note-array";

export enum ModeEnum {
  NaturalMajor,
  HarmonicMajor,
  MelodicMajor,

  NaturalMinor,
  HarmonicMinor,
  MelodicMinor,

  Ionian,
  Dorian,
  Phrygian,
  Lydian,
  Mixolydian,
  Aeolian,
  Locrian,
}

const majorKeys = [
  "C",
  "G",
  "D",
  "A",
  "E",
  "B",
  "F",
  "B♭",
  "E♭",
  "A♭",
  "D♭",
  "G♭",
];

const minorKeys = [
  "A",
  "E",
  "B",
  "F♯",
  "C♯",
  "G♯",
  "D",
  "G",
  "C",
  "F",
  "B♭",
  "D♯",
];

const modeEnumToKeys: Record<ModeEnum, string[]> = {
  [ModeEnum.NaturalMajor]: majorKeys,
  [ModeEnum.HarmonicMajor]: majorKeys,
  [ModeEnum.MelodicMajor]: majorKeys,

  [ModeEnum.NaturalMinor]: minorKeys,
  [ModeEnum.HarmonicMinor]: minorKeys,
  [ModeEnum.MelodicMinor]: minorKeys,

  // TODO:
  [ModeEnum.Ionian]: majorKeys,
  [ModeEnum.Dorian]: [],
  [ModeEnum.Phrygian]: [],
  [ModeEnum.Lydian]: [],
  [ModeEnum.Mixolydian]: [],
  [ModeEnum.Aeolian]: minorKeys,
  [ModeEnum.Locrian]: [],
};

const majorMinorRelations: Record<number, ModeEnum> = {
  [ModeEnum.NaturalMajor]: ModeEnum.NaturalMinor,
  [ModeEnum.HarmonicMajor]: ModeEnum.HarmonicMinor,
  [ModeEnum.MelodicMajor]: ModeEnum.MelodicMinor,

  [ModeEnum.NaturalMinor]: ModeEnum.NaturalMajor,
  [ModeEnum.HarmonicMinor]: ModeEnum.HarmonicMajor,
  [ModeEnum.MelodicMinor]: ModeEnum.MelodicMajor,
};

const modeEnumToString: Record<ModeEnum, string> = {
  [ModeEnum.NaturalMajor]: "Natural Major",
  [ModeEnum.HarmonicMajor]: "Harmonic Major",
  [ModeEnum.MelodicMajor]: "Melodic Major",

  [ModeEnum.NaturalMinor]: "Natural Minor",
  [ModeEnum.HarmonicMinor]: "Harmonic Minor",
  [ModeEnum.MelodicMinor]: "Melodic Minor",

  [ModeEnum.Ionian]: "Ionian", // same as ModeEnum.NaturalMajor
  [ModeEnum.Dorian]: "Dorian",
  [ModeEnum.Phrygian]: "Phrygian",
  [ModeEnum.Lydian]: "Lydian",
  [ModeEnum.Mixolydian]: "Mixolydian",
  [ModeEnum.Aeolian]: "Aeolian", // same as ModeEnum.NaturalMinor
  [ModeEnum.Locrian]: "Locrian",
};

const chordTypeEnumToIndexes: Record<ChordTypeEnum, number[]> = {
  [ChordTypeEnum.Triad]: [1, 3, 5],
  [ChordTypeEnum.Seventh]: [1, 3, 5, 7],
  // [ChordTypeEnum.Ninth]: [1, 3, 5, 7, 9],
  // [ChordTypeEnum.Eleventh]: [1, 3, 5, 7, 9, 11],
  // [ChordTypeEnum.Thirteen]: [1, 3, 5, 7, 9, 11, 13],
};

const modeToIntervals = transformObject(
  {
    [ModeEnum.NaturalMajor]: ["P1", "M2", "M3", "P4", "P5", "M6", "M7"],
    [ModeEnum.HarmonicMajor]: ["P1", "M2", "M3", "P4", "P5", "m6", "M7"],
    [ModeEnum.MelodicMajor]: ["P1", "M2", "M3", "P4", "P5", "m6", "m7"],

    [ModeEnum.NaturalMinor]: ["P1", "M2", "m3", "P4", "P5", "m6", "m7"],
    [ModeEnum.HarmonicMinor]: ["P1", "M2", "m3", "P4", "P5", "m6", "M7"],
    [ModeEnum.MelodicMinor]: ["P1", "M2", "m3", "P4", "P5", "M6", "M7"],

    [ModeEnum.Ionian]: ["P1", "M2", "M3", "P4", "P5", "M6", "M7"], // same as ModeEnum.NaturalMajor
    [ModeEnum.Dorian]: ["P1", "M2", "m3", "P4", "P5", "M6", "m7"],
    [ModeEnum.Phrygian]: ["P1", "m2", "m3", "P4", "P5", "m6", "m7"],
    [ModeEnum.Lydian]: ["P1", "M2", "M3", "A4", "P5", "M6", "M7"],
    [ModeEnum.Mixolydian]: ["P1", "M2", "M3", "P4", "P5", "M6", "m7"],
    [ModeEnum.Aeolian]: ["P1", "M2", "m3", "P4", "P5", "m6", "m7"], // same as ModeEnum.NaturalMinor
    [ModeEnum.Locrian]: ["P1", "m2", "m3", "P4", "d5", "m6", "m7"],
  },
  (_, abbrs) => abbrs.map((abbr) => Interval.from(abbr))
);

type ModeNameOptions = { shortName?: boolean } & NoteToNameOptions;

export class Mode {
  private _noteArr: NoteArray;

  private _mode?: ModeEnum;

  constructor(notes: Notes | NoteArray, mode?: ModeEnum) {
    this._noteArr = Array.isArray(notes) ? new NoteArray(notes) : notes;

    if (this._noteArr.count() === 0) {
      throw new Error("mode must contains at least one note");
    }

    this._mode = mode;
  }

  static from(key: NoteType, mode: ModeEnum) {
    const intervals = modeToIntervals[mode];

    if (!intervals) {
      throw new Error(`unknown mode: ${mode}`);
    }

    const root = typeof key === "string" ? Note.from(key) : key;
    const notes = intervals.map((inter) => root.add(inter));

    // In harmonic minor, force set 7th note(which accidental is AccidentalEnum.Natural) to AccidentalEnum.NaturalWithSymbol.
    if (
      mode === ModeEnum.HarmonicMinor &&
      notes[6].accidental === AccidentalEnum.Natural
    ) {
      notes[6].accidental = AccidentalEnum.NaturalWithSymbol;
    }

    return new Mode(notes, mode);
  }

  static modes() {
    return Object.entries(modeEnumToString).map(([key, value]) => ({
      label: value,
      value: Number(key) as ModeEnum,
    }));
  }

  static getKeys(mode: ModeEnum) {
    const keys = modeEnumToKeys[mode];

    if (!keys) {
      throw new Error(`unsupported mode: ${mode}`);
    }

    return keys;
  }

  key() {
    return this._noteArr.get(0)!;
  }

  type() {
    return this._mode;
  }

  chord(step: number, type?: ChordTypeEnum) {
    if (step < 1 || step > 7) {
      throw new Error("step must be between 1 and 7");
    }

    const indexes = chordTypeEnumToIndexes[type || ChordTypeEnum.Triad];

    if (!indexes) {
      throw new Error(`unknown chord type: ${type}`);
    }

    const notes = indexes.map((index) => {
      const idxAbs = step + index - 1;

      let idx = idxAbs % 7;
      idx = idx < 1 ? idx + 7 : idx;

      const note = this._noteArr.get(idx - 1)!;

      if (note.group) {
        note.group += Math.floor((idxAbs - 1) / 7);
      }

      return note;
    });

    return new Chord(notes);
  }

  transpose(direction: "low" | "high", inter: Inter) {
    if (direction === "low") {
      this._noteArr = this._noteArr.map((note) => note.minus(inter));
    }
    if (direction === "high") {
      this._noteArr = this._noteArr.map((note) => note.add(inter));
    }

    return this;
  }

  parallel() {
    if (!this.isMajor() && !this.isMinor()) {
      throw new Error("parallel() only works on major or minor mode");
    }

    const root = this._noteArr.root();
    const mode = majorMinorRelations[this._mode!];

    return Mode.from(root, mode);
  }

  relative() {
    if (!this.isMajor() && !this.isMinor()) {
      throw new Error("relative() only works on major or minor mode");
    }

    let root = this._noteArr.root();

    if (this.isMajor()) {
      root = root.minus("m3");
    }
    if (this.isMinor()) {
      root = root.add("m3");
    }

    const mode = majorMinorRelations[this._mode!];

    return Mode.from(root, mode);
  }

  name(opts?: ModeNameOptions) {
    const root = this._noteArr.root().name(opts);

    let mode =
      typeof this._mode === "undefined"
        ? "Unknown Mode"
        : modeEnumToString[this._mode];

    if (opts?.shortName) {
      mode = mode
        .replace("Natural", "")
        .replace("Harmonic", "")
        .replace("Melodic", "");
    }

    return `${root} ${mode}`;
  }

  notes() {
    return this._noteArr;
  }

  clone() {
    return new Mode(this._noteArr.clone(), this._mode);
  }

  private isMajor() {
    return (
      typeof this._mode !== "undefined" &&
      [
        ModeEnum.NaturalMajor,
        ModeEnum.HarmonicMajor,
        ModeEnum.MelodicMajor,
      ].includes(this._mode)
    );
  }

  private isMinor() {
    return (
      typeof this._mode !== "undefined" &&
      [
        ModeEnum.NaturalMinor,
        ModeEnum.HarmonicMinor,
        ModeEnum.MelodicMinor,
      ].includes(this._mode)
    );
  }
}
