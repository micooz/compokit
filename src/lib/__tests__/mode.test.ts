import { ChordTypeEnum } from "../chord";
import { Mode, ModeEnum } from "../mode";
import { Note } from "../note";

test("scale notes should match", () => {
  const scales = [
    // majors
    ...["C", "G", "D", "A", "E", "B", "Gb", "Db", "Ab", "Eb", "Bb", "F"].map(
      (key) => Mode.from(Note.from(key), ModeEnum.NaturalMajor)
    ),
    // minors
    ...["A", "E", "B", "F#", "C#", "G#", "D#", "Bb", "F", "C", "G", "D"].map(
      (key) => Mode.from(Note.from(key), ModeEnum.NaturalMinor)
    ),
    ...["A", "E", "B", "F#", "C#", "G#", "D#", "Bb", "F", "C", "G", "D"].map(
      (key) => Mode.from(Note.from(key), ModeEnum.HarmonicMinor)
    ),
  ];

  const lines = scales.map(
    (scale) =>
      `${scale.name({ transformAccidental: true })}: ${scale
        .notes()
        .join(" ", { transformAccidental: true })}`
  );

  expect(lines).toMatchSnapshot();
});

test("scale note with group should match", () => {
  const scale = Mode.from("Ab3", ModeEnum.NaturalMajor);

  expect(scale.notes().join()).toBe("Ab3 Bb3 C4 Db4 Eb4 F4 G4");
});

test("unknown mode should throw", () => {
  // @ts-expect-error
  expect(() => Mode.from("C3", -1)).toThrow();
});

test("Mode:modes", () => {
  expect(Mode.modes()).toMatchSnapshot();
});

test("Mode:getKeys", () => {
  expect(Mode.getKeys(ModeEnum.NaturalMajor)).toMatchSnapshot();
  expect(() => Mode.getKeys(-1 as any)).toThrow("unsupported mode: -1");
});

test("Mode::chord", () => {
  const mode = Mode.from("Eb", ModeEnum.NaturalMajor);

  expect(() => mode.chord(0, ChordTypeEnum.Triad)).toThrow(
    "step must be between 1 and 7"
  );
  expect(() => mode.chord(1, -1 as any)).toThrow("unknown chord type: -1");

  expect(mode.chord(1).join()).toBe("Eb G Bb");
  expect(mode.chord(2, ChordTypeEnum.Seventh).join()).toBe("F Ab C Eb");
  // expect(mode.chord(3, ChordTypeEnum.Ninth).join()).toBe(
  //   "G Bb D F Ab"
  // );
  // expect(mode.chord(4, ChordTypeEnum.Eleventh).join()).toBe(
  //   "Ab C Eb G Bb D"
  // );

  // with group
  const mode2 = Mode.from("C4", ModeEnum.NaturalMajor);
  expect(mode2.chord(5).join()).toBe("G4 B4 D5");
});

test("Mode::transpose", () => {
  expect(
    Mode.from("Eb", ModeEnum.NaturalMajor).transpose("low", "m3").notes().join()
  ).toBe("C D E F G A B");

  expect(
    Mode.from("A", ModeEnum.NaturalMinor).transpose("high", "m3").notes().join()
  ).toBe("C D Eb F G Ab Bb");
});

test("Mode::parallel", () => {
  expect(Mode.from("C", ModeEnum.NaturalMajor).parallel().notes().join()).toBe(
    "C D Eb F G Ab Bb"
  );

  expect(Mode.from("C", ModeEnum.HarmonicMinor).parallel().notes().join()).toBe(
    "C D E F G Ab B"
  );

  expect(() => new Mode([]).parallel()).toThrow(
    "parallel() only works on major or minor mode"
  );
});

test("Mode::relative", () => {
  expect(Mode.from("C", ModeEnum.NaturalMajor).relative().notes().join()).toBe(
    "A B C D E F G"
  );

  expect(Mode.from("A", ModeEnum.HarmonicMinor).relative().notes().join()).toBe(
    "C D E F G Ab B"
  );

  expect(() => new Mode([]).relative()).toThrow(
    "relative() only works on major or minor mode"
  );
});

test("Mode::name", () => {
  const mode = new Mode([Note.from("C"), Note.from("Eb")]);
  expect(mode.name()).toBe("C Unknown Mode");
});

test("Mode::notes", () => {
  const mode = new Mode([Note.from("C"), Note.from("Eb")]);
  expect(mode.notes().join()).toBe("C Eb");
});
