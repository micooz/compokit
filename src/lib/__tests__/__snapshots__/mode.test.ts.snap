// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Mode:getKeys 1`] = `
[
  "C",
  "G",
  "D",
  "A",
  "E",
  "B",
  "F",
  "B♭",
  "E♭",
  "A♭",
  "D♭",
  "G♭",
]
`;

exports[`Mode:modes 1`] = `
[
  {
    "label": "Natural Major",
    "value": 0,
  },
  {
    "label": "Harmonic Major",
    "value": 1,
  },
  {
    "label": "Melodic Major",
    "value": 2,
  },
  {
    "label": "Natural Minor",
    "value": 3,
  },
  {
    "label": "Harmonic Minor",
    "value": 4,
  },
  {
    "label": "Melodic Minor",
    "value": 5,
  },
  {
    "label": "Ionian",
    "value": 6,
  },
  {
    "label": "Dorian",
    "value": 7,
  },
  {
    "label": "Phrygian",
    "value": 8,
  },
  {
    "label": "Lydian",
    "value": 9,
  },
  {
    "label": "Mixolydian",
    "value": 10,
  },
  {
    "label": "Aeolian",
    "value": 11,
  },
  {
    "label": "Locrian",
    "value": 12,
  },
]
`;

exports[`scale notes should match 1`] = `
[
  "C Natural Major: C D E F G A B",
  "G Natural Major: G A B C D E F♯",
  "D Natural Major: D E F♯ G A B C♯",
  "A Natural Major: A B C♯ D E F♯ G♯",
  "E Natural Major: E F♯ G♯ A B C♯ D♯",
  "B Natural Major: B C♯ D♯ E F♯ G♯ A♯",
  "G♭ Natural Major: G♭ A♭ B♭ C♭ D♭ E♭ F",
  "D♭ Natural Major: D♭ E♭ F G♭ A♭ B♭ C",
  "A♭ Natural Major: A♭ B♭ C D♭ E♭ F G",
  "E♭ Natural Major: E♭ F G A♭ B♭ C D",
  "B♭ Natural Major: B♭ C D E♭ F G A",
  "F Natural Major: F G A B♭ C D E",
  "A Natural Minor: A B C D E F G",
  "E Natural Minor: E F♯ G A B C D",
  "B Natural Minor: B C♯ D E F♯ G A",
  "F♯ Natural Minor: F♯ G♯ A B C♯ D E",
  "C♯ Natural Minor: C♯ D♯ E F♯ G♯ A B",
  "G♯ Natural Minor: G♯ A♯ B C♯ D♯ E F♯",
  "D♯ Natural Minor: D♯ E♯ F♯ G♯ A♯ B C♯",
  "B♭ Natural Minor: B♭ C D♭ E♭ F G♭ A♭",
  "F Natural Minor: F G A♭ B♭ C D♭ E♭",
  "C Natural Minor: C D E♭ F G A♭ B♭",
  "G Natural Minor: G A B♭ C D E♭ F",
  "D Natural Minor: D E F G A B♭ C",
  "A Harmonic Minor: A B C D E F G♯",
  "E Harmonic Minor: E F♯ G A B C D♯",
  "B Harmonic Minor: B C♯ D E F♯ G A♯",
  "F♯ Harmonic Minor: F♯ G♯ A B C♯ D E♯",
  "C♯ Harmonic Minor: C♯ D♯ E F♯ G♯ A B♯",
  "G♯ Harmonic Minor: G♯ A♯ B C♯ D♯ E F𝄪",
  "D♯ Harmonic Minor: D♯ E♯ F♯ G♯ A♯ B C𝄪",
  "B♭ Harmonic Minor: B♭ C D♭ E♭ F G♭ A♮",
  "F Harmonic Minor: F G A♭ B♭ C D♭ E♮",
  "C Harmonic Minor: C D E♭ F G A♭ B♮",
  "G Harmonic Minor: G A B♭ C D E♭ F♯",
  "D Harmonic Minor: D E F G A B♭ C♯",
]
`;
